File: js/test/entity/webhook/WebhookEntity.test.js

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

const Path = require('node:path')
const Fs = require('node:fs')

const { test, describe } = require('node:test')
const assert = require('node:assert')


const { BluefinPayconexSDK, BaseFeature, stdutil, config } = require('../../..')

const {
  envOverride,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
} = require('../../utility')


describe('WebhookEntity', async () => {

  test('instance', async () => {
    const testsdk = BluefinPayconexSDK.test()
    const ent = testsdk.Webhook()
    assert(null != ent)
  })


  // Feature #4: the entity `stream(action, ...)` method runs the op pipeline
  // and returns an async iterator over result items. With the streaming
  // feature active it yields the feature's incremental output; otherwise it
  // falls back to the materialised list so `stream` always yields.
  test('stream', async () => {
    const seed = {
      entity: {
        webhook: { s1: { id: 's1' }, s2: { id: 's2' }, s3: { id: 's3' } }
      }
    }

    // Fallback: streaming inactive -> yields the materialised list items.
    const base = BluefinPayconexSDK.test(seed)
    const seen = []
    for await (const item of base.Webhook().stream('list')) {
      seen.push(item)
    }
    assert.equal(seen.length, 3)

    // Inbound: streaming active -> yields each item from the feature iterator.
    if (config.feature && config.feature.streaming) {
      const sdk = BluefinPayconexSDK.test(seed, { feature: { streaming: { active: true } } })
      const got = []
      for await (const item of sdk.Webhook().stream('list')) {
        if (Array.isArray(item)) { got.push(...item) } else { got.push(item) }
      }
      assert.equal(got.length, 3)
    }
  })


  test('basic', async () => {

    const setup = basicSetup()
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const webhook_ref01_ent = client.Webhook()
    let webhook_ref01_data = setup.data.new.webhook['webhook_ref01']
    webhook_ref01_data['account_id'] = setup.idmap['account01']

    webhook_ref01_data = await webhook_ref01_ent.create(webhook_ref01_data)
    assert(null != webhook_ref01_data.id)


    // LIST
    const webhook_ref01_match = {}
    webhook_ref01_match['account_id'] = setup.idmap['account01']

    const webhook_ref01_list = await webhook_ref01_ent.list(webhook_ref01_match)

    assert(!isempty(select(webhook_ref01_list, { id: webhook_ref01_data.id })))


    // UPDATE
    const webhook_ref01_data_up0 = {}
    webhook_ref01_data_up0.id = webhook_ref01_data.id
    webhook_ref01_data_up0 ['account_id'] = setup.idmap['account_id']

    const webhook_ref01_markdef_up0 = { name: 'event', value: 'Mark01-webhook_ref01_' + setup.now }
    webhook_ref01_data_up0 [webhook_ref01_markdef_up0.name] = webhook_ref01_markdef_up0.value

    const webhook_ref01_resdata_up0 = await webhook_ref01_ent.update(webhook_ref01_data_up0)
    assert(webhook_ref01_resdata_up0.id === webhook_ref01_data_up0.id)

    assert(webhook_ref01_resdata_up0[webhook_ref01_markdef_up0.name] === webhook_ref01_markdef_up0.value)


    // LOAD
    const webhook_ref01_match_dt0 = {}
    webhook_ref01_match_dt0.id = webhook_ref01_data.id
    const webhook_ref01_data_dt0 = await webhook_ref01_ent.load(webhook_ref01_match_dt0)
    assert(webhook_ref01_data_dt0.id === webhook_ref01_data.id)


    // REMOVE
    const webhook_ref01_match_rm0 = {}
    webhook_ref01_match_rm0.id = webhook_ref01_data.id
    await webhook_ref01_ent.remove(webhook_ref01_match_rm0)
  

    // LIST
    const webhook_ref01_match_rt0 = {}
    webhook_ref01_match_rt0['account_id'] = setup.idmap['account01']

    const webhook_ref01_list_rt0 = await webhook_ref01_ent.list(webhook_ref01_match_rt0)

    assert(isempty(select(webhook_ref01_list_rt0, { id: webhook_ref01_data.id })))


  })
})



function basicSetup(extra) {
  // TODO: fix test def options
  const options = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname,
      '../../../../.sdk/test/entity/webhook/WebhookTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = BluefinPayconexSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['webhook01','webhook02','webhook03','account01','account02','account03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'BLUEFIN_PAYCONEX_TEST_WEBHOOK_ENTID': idmap,
    'BLUEFIN_PAYCONEX_TEST_LIVE': 'FALSE',
    'BLUEFIN_PAYCONEX_TEST_EXPLAIN': 'FALSE',
    'BLUEFIN_PAYCONEX_APIKEY': 'NONE',
  })

  idmap = env['BLUEFIN_PAYCONEX_TEST_WEBHOOK_ENTID']

  if ('TRUE' === env.BLUEFIN_PAYCONEX_TEST_LIVE) {
    client = new BluefinPayconexSDK(merge([
      {
        apikey: env.BLUEFIN_PAYCONEX_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.BLUEFIN_PAYCONEX_TEST_EXPLAIN,
    now: Date.now(),
  }

  return setup
}
  

File: scala/sdktest/WebhookEntityTest.scala
// Generated basic-flow test for the webhook entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped WebhookTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefinpayconexsdk.core.{Helpers, BluefinPayconexSDK}
import voxgig.bluefinpayconexsdk.utility.struct.Struct

object WebhookEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("webhook.instance") {
      val testsdk = BluefinPayconexSDK.testSDK()
      val ent = testsdk.webhook(null)
      rep.check("webhook.instance", ent != null, "expected non-null webhook entity")
    }

    rep.scope("webhook.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/webhook/WebhookTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinPayconexSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("webhook01", "WEBHOOK01")
      idmap.put("webhook02", "WEBHOOK02")
      idmap.put("webhook03", "WEBHOOK03")
      idmap.put("account01", "ACCOUNT01")
      idmap.put("account02", "ACCOUNT02")
      idmap.put("account03", "ACCOUNT03")
      idmap.put("account_id", "ACCOUNT01")
      val now = System.currentTimeMillis()

      // CREATE
      val webhookRef01Ent = client.webhook(null)
      var webhookRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.webhook"), "webhook_ref01"))
      webhookRef01Data.put("account_id", idmap.get("account01"))
      val webhookRef01DataResult = webhookRef01Ent.create(webhookRef01Data, null)
      webhookRef01Data = Helpers.toMapAny(webhookRef01DataResult)
      rep.check("webhook.create.map", webhookRef01Data != null, "expected create result to be a map")
      rep.check("webhook.create.id", webhookRef01Data != null && webhookRef01Data.get("id") != null, "expected created entity to have an id")

      // LIST
      val webhookRef01Match = new LinkedHashMap[String, Object]()
      webhookRef01Match.put("account_id", idmap.get("account01"))
      val webhookRef01ListResult = webhookRef01Ent.list(webhookRef01Match, null)
      rep.check("webhook.list.islist", webhookRef01ListResult.isInstanceOf[JList[?]], "expected list result to be an array, got " + webhookRef01ListResult)
      val webhookRef01List = webhookRef01ListResult.asInstanceOf[JList[Object]]
      val webhookRef01ListFound = Struct.select(
          SdkTestSupport.entityListToData(webhookRef01List), SdkTestSupport.om("id" -> webhookRef01Data.get("id")))
      rep.check("webhook.list.exists", !Struct.isempty(webhookRef01ListFound), "expected to find created entity in list")

      // UPDATE
      val webhookRef01DataUp0Up = new LinkedHashMap[String, Object]()
      webhookRef01DataUp0Up.put("id", webhookRef01Data.get("id"))
      webhookRef01DataUp0Up.put("account_id", idmap.get("account_id"))
      val webhookRef01MarkdefUp0Name = "event"
      val webhookRef01MarkdefUp0Value = "Mark01-webhook_ref01_" + now
      webhookRef01DataUp0Up.put(webhookRef01MarkdefUp0Name, webhookRef01MarkdefUp0Value)
      val webhookRef01ResdataUp0Result = webhookRef01Ent.update(webhookRef01DataUp0Up, null)
      val webhookRef01ResdataUp0 = Helpers.toMapAny(webhookRef01ResdataUp0Result)
      rep.check("webhook.update.map", webhookRef01ResdataUp0 != null, "expected update result to be a map")
      rep.eq("webhook.update.id", webhookRef01DataUp0Up.get("id"), webhookRef01ResdataUp0.get("id"))
      rep.eq("webhook.update.mark", webhookRef01MarkdefUp0Value, webhookRef01ResdataUp0.get(webhookRef01MarkdefUp0Name))

      // LOAD
      val webhookRef01MatchDt0 = new LinkedHashMap[String, Object]()
      webhookRef01MatchDt0.put("id", webhookRef01Data.get("id"))
      val webhookRef01DataDt0Loaded = webhookRef01Ent.load(webhookRef01MatchDt0, null)
      val webhookRef01DataDt0LoadResult = Helpers.toMapAny(webhookRef01DataDt0Loaded)
      rep.check("webhook.load.map", webhookRef01DataDt0LoadResult != null, "expected load result to be a map")
      rep.eq("webhook.load.id", webhookRef01Data.get("id"), webhookRef01DataDt0LoadResult.get("id"))

      // REMOVE
      val webhookRef01MatchRm0 = new LinkedHashMap[String, Object]()
      webhookRef01MatchRm0.put("id", webhookRef01Data.get("id"))
      webhookRef01Ent.remove(webhookRef01MatchRm0, null)

      // LIST
      val webhookRef01MatchRt0 = new LinkedHashMap[String, Object]()
      webhookRef01MatchRt0.put("account_id", idmap.get("account01"))
      val webhookRef01ListRt0Result = webhookRef01Ent.list(webhookRef01MatchRt0, null)
      rep.check("webhook.list.islist", webhookRef01ListRt0Result.isInstanceOf[JList[?]], "expected list result to be an array, got " + webhookRef01ListRt0Result)
      val webhookRef01ListRt0 = webhookRef01ListRt0Result.asInstanceOf[JList[Object]]
      val webhookRef01ListRt0NotFound = Struct.select(
          SdkTestSupport.entityListToData(webhookRef01ListRt0), SdkTestSupport.om("id" -> webhookRef01Data.get("id")))
      rep.check("webhook.list.notexists", Struct.isempty(webhookRef01ListRt0NotFound), "expected removed entity to not be in list")
    }
  }
}
